File: 22.generate-parentheses.kt
// @leet start
class Solution {
    fun backtrack(res: MutableList<String>, cur: String, left: Int, right: Int) {
      if (left == 0 && right == 0) res.add(cur)
      if (left > 0) backtrack(res, "$cur(", left - 1, right)
      if (right > left) backtrack(res, "$cur)", left, right - 1)
    }

    fun generateParenthesis(n: Int): List<String> {
      var res = mutableListOf<String>()
      backtrack(res, "", n, n)
      return res
    }
}
// @leet end

File: 22.generate-parentheses.cs
// @leet start
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
      var res = new List<string>();
      Action<int,int,string> backtrack = null;
      backtrack = (left, right, cur) => {
        if (left == n && right == n) res.Add(cur);
        if (left < n) backtrack(left + 1, right, cur + "(");
        if (right < left) backtrack(left, right + 1, cur + ")");
      };
      backtrack(0, 0, "");
      return res;
    }
}
// @leet end
